File: web-ui/src/lib/api_types.ts
// Responses

type LibrariesResponse = ApiLibraryEntry[];

type FileInfoResponse = (ApiFileInfo & { type: "file" }) | (ApiDirectoryInfo & { type: "directory" });

interface ListDirectoryResponse {
	files: ApiFileEntry[],
	directories: ApiDirectoryEntry[],
	total_duration: number,
	total_size: number,
}

interface ApiWatchHistoryResponse {
	total_pages: number,
	entries: ApiWatchHistoryEntry[],
}

// Params

interface UpdateWatchProgressParams {
	library_id: string,
	media_path: string,
	new_watch_progress: number,
}

interface DeleteWatchProgressParams {
	library_id: string,
	media_path: string,
}

// Shared Types

interface ApiUserInfo {
	display_name: string,
	username: string,
}

interface ApiLibraryEntry {
	id: string,
	display_name: string,
}

interface ApiFileEntry {
	path_name: string,
	full_path: string,
	display_name: string,
	thumbnail_path: string,
	duration: number,
	file_size: number,
	artist: string | null,
	watch_progress: number | null,
	creation_date: string,
}

interface ApiDirectoryEntry {
	path_name: string,
	display_name: string,
	thumbnail_path: string | null,
	child_count: number,
}

interface ApiInfoCommon {
	full_path: string,
	library_display_name: string,
	display_name: string,
}

interface ApiFileInfo extends ApiInfoCommon {
	file_size: number,
	duration: number,
	artist: string | null,
	creation_date: string,
	video_info: ApiVideoInfo | null,
	prev_video: string | null,
	next_video: string | null,
	watch_progress: number | null,
	description: string | null,
	connections: ApiVideoConnection[],
	comments: ApiCommentThread[],
}

interface ApiVideoInfo {
	video_size: ApiDimension,
	sheet_thumbnail_size: ApiDimension,
	thumbnail_sheet_rows: number,
	thumbnail_sheet_cols: number,
	thumbnail_sheet_interval: number,
}

interface ApiVideoConnection {
	video_path: string,
	video_thumbnail: string,
	relation: string,
	shortcut_thumbnail: string | null,
	left_start: number,
	left_end: number,
	right_start: number,
}

interface ApiCommentThread {
	comment: ApiComment,
	replies: ApiComment[],
}

interface ApiComment {
	author: string,
	text: string,
	likes: number,
	published_at: string,
}

interface ApiDirectoryInfo extends ApiInfoCommon {}

interface ApiWatchHistoryEntry {
	library_id: string,
	media_path: string,
	last_watched: string,
	progress: number,
	file: ApiFileEntry | null,
}

interface ApiDimension {
	width: number,
	height: number,
}

File: src/web_server/api_types.rs
use relative_path::RelativePathBuf;
use serde::{Deserialize, Serialize};
use time::OffsetDateTime;

use crate::web_server::media_metadata::Dimension;

#[derive(Debug, Serialize)]
pub struct ApiUserInfo {
	pub display_name: String,
	pub username: String,
}

#[derive(Debug, Serialize)]
pub struct ApiLibraryEntry {
	pub id: String,
	pub display_name: String,
}

#[derive(Debug, Serialize)]
pub struct ApiFileEntry {
	pub path_name: String,
	pub full_path: RelativePathBuf,
	pub display_name: String,
	pub thumbnail_path: String,
	pub duration: u64,
	pub file_size: u64,
	pub artist: Option<String>,
	pub watch_progress: Option<u64>,
	#[serde(with = "time::serde::iso8601")]
	pub creation_date: OffsetDateTime,
}

#[derive(Debug, Serialize)]
pub struct ApiDirectoryEntry {
	pub path_name: String,
	pub display_name: String,
	pub thumbnail_path: Option<String>,
	pub child_count: u32,
}

#[derive(Debug, Serialize)]
pub struct ApiFileInfo {
	pub full_path: RelativePathBuf,
	pub library_display_name: String,
	pub display_name: String,
	pub file_size: u64,
	pub duration: u64,
	pub artist: Option<String>,
	#[serde(with = "time::serde::iso8601")]
	pub creation_date: OffsetDateTime,
	pub thumbnail_path: String,
	pub video_info: Option<ApiVideoInfo>,
	pub subtitle_streams: Vec<ApiSubtitleStream>,
	pub prev_video: Option<String>,
	pub next_video: Option<String>,
	pub watch_progress: Option<u64>,
	pub description: Option<String>,
	pub connections: Vec<ApiVideoConnection>,
	pub comments: Vec<ApiCommentThread>,
}

#[derive(Debug, Serialize)]
pub struct ApiVideoInfo {
	pub video_size: Dimension,
	pub sheet_thumbnail_size: Dimension,
	pub thumbnail_sheet_rows: u32,
	pub thumbnail_sheet_cols: u32,
	pub thumbnail_sheet_interval: u32,
}

#[derive(Debug, Serialize)]
pub struct ApiVideoConnection {
	pub video_path: RelativePathBuf,
	pub video_thumbnail: String,
	pub relation: String,
	pub shortcut_thumbnail: Option<String>,
	pub left_start: u64,
	pub left_end: u64,
	pub right_start: u64,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ApiSubtitleStream {
	pub index: usize,
	pub language: Option<String>,
	pub name: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ApiCommentThread {
	pub comment: ApiComment,
	pub replies: Vec<ApiComment>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ApiComment {
	pub author: String,
	pub text: String,
	pub likes: u64,
	#[serde(with = "time::serde::iso8601")]
	pub published_at: OffsetDateTime,
}

#[derive(Debug, Serialize)]
pub struct ApiDirectoryInfo {
	pub full_path: RelativePathBuf,
	pub library_display_name: String,
	pub display_name: String,
}

#[derive(Debug, Serialize)]
pub struct ApiWatchHistoryEntry {
	pub library_id: String,
	pub media_path: RelativePathBuf,
	#[serde(with = "time::serde::iso8601")]
	pub last_watched: OffsetDateTime,
	pub progress: u64,
	pub file: Option<ApiFileEntry>,
}
